[GStreamer] Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in FastMalloc allocator

philn opened the pull request:
#### 198b50a5a3a8581bcfa48f619e7033d2cdc3dbad
<pre>
[GStreamer] Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in FastMalloc allocator
<a href="https://bugs.webkit.org/show_bug.cgi?id=285528">https://bugs.webkit.org/show_bug.cgi?id=285528</a>

Reviewed by Xabier Rodriguez-Calvar.

The allocator was updated to modern WebKit/GObject code style and now stores allocated data in a
span.

Co-authored-by: Alicia Boya Garcia &lt;aboya@igalia.com&gt;

* Source/WTF/wtf/MallocSpan.h:
(WTF::MallocSpan::tryAlignedMalloc):
* Source/WebCore/platform/graphics/gstreamer/GstAllocatorFastMalloc.cpp:
(gstMemoryFastMallocNew):
(gstAllocatorFastMallocAlloc):
(gstAllocatorFastMallocMemMap):
(gstAllocatorFastMallocMemUnmap):
(gstAllocatorFastMallocMemCopy):
(gstAllocatorFastMallocMemShare):
(gstAllocatorFastMallocMemIsSpan):
(gstAllocatorFastMallocConstructed):
(gst_allocator_fast_malloc_class_init):
(gst_allocator_fast_malloc_init): Deleted.
* Tools/Scripts/webkitpy/style/checker.py:

Canonical link: <a href="https://commits.webkit.org/288928@main">https://commits.webkit.org/288928@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d342e5308563d9968c656d48f2d5076cd2465bd2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84177 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3798 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38481 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89253 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35186 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86262 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3886 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11787 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65475 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23312 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87223 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2913 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76492 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45768 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/83586 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2861 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30722 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34235 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/77141 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73785 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31486 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90634 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/83194 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11444 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8320 "Found 2 new test failures: fast/selectors/selection-window-inactive-stroke-color.html imported/w3c/web-platform-tests/webcodecs/audio-encoder.https.any.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73931 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11670 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72317 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73135 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18207 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17441 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2796 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11396 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16872 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/105613 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11245 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25507 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14721 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13018 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/GstAllocatorFastMalloc.cpp:
```diff
@@ -21,24 +21,31 @@
-G_DEFINE_TYPE(GstAllocatorFastMalloc, gst_allocator_fast_malloc, GST_TYPE_ALLOCATOR)
-WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
+WEBKIT_DEFINE_TYPE(GstAllocatorFastMalloc, gst_allocator_fast_malloc, GST_TYPE_ALLOCATOR)
+#define parent_class gst_allocator_fast_malloc_parent_class
```
I'll rework this a bit, we should get rid of GST_CALL_PARENT macros.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/GstAllocatorFastMalloc.cpp:
```diff
@@ -52,21 +58,21 @@ static GstMemoryFastMalloc* gstMemoryFastMallocNew(GstAllocator* allocator, gsiz
         return nullptr;
 
-    memory->data = reinterpret_cast<uint8_t*>(memory) + headerSize;
+    auto data = unsafeMakeSpan(reinterpret_cast<uint8_t*>(memory), totalSize);
```
Instead of `unsafeMakeSpan` we could use `MallocSpan<uint8_t>::tryAlignedMalloc()` which would return already the span... if such variant existed!

`MallocSpan<T>::alignedMalloc()` exists, and so does `MallocSpan<T>::tryZeroedMalloc()`, so I see no reason we couldn't add `tryAlignedMalloc()` for completion.

ntrrgc replied:
We could be using more significative names. Clarity helps when dealing with already pretty complicated code.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/GstAllocatorFastMalloc.cpp:
```diff
@@ -52,21 +58,21 @@ static GstMemoryFastMalloc* gstMemoryFastMallocNew(GstAllocator* allocator, gsiz
+    auto data = unsafeMakeSpan(reinterpret_cast<uint8_t*>(memory), totalSize);
+    memory->data = data.subspan(headerSize);
 
+    auto span = memory->data.subspan(0, size);
```
This will fail 3 lines below if padding > 0. The size of the tail is `offset + size + padding`.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/GstAllocatorFastMalloc.cpp:
```diff
@@ -129,39 +137,45 @@ static GstMemoryFastMalloc* gstAllocatorFastMallocMemShare(GstMemoryFastMalloc*
 }
 
-static gboolean gstAllocatorFastMallocMemIsSpan(GstMemoryFastMalloc* memory, GstMemoryFastMalloc* other, gsize* offset)
+static gboolean gstAllocatorFastMallocMemIsSpan(GstMemory* baseMemory, GstMemory* otherMemory, gsize* offset)
```
`baseMemory` is not a good name here. It could be misunderstood to mean it is the parent memory or something, which is not the case.

gstAllocatorFastMallocMemIsSpan() will be called by gst_memory_is_span() after checking that both GstMemories have the same allocators and they both share the same non-null parent.
The function must return whether the visible end of the first memory coincides with the visible start of the second memory and fill `offset` with the offset of the first memory within the visible span of the parent memory.

This is used to avoid expensive concatenation: If the function returns true, it means you can use gst_memory_share() on the parent memory with the filled offset and the combined size of both memories to get a GstMemory that spans both previous memories without doing any copies.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/GstAllocatorFastMalloc.cpp:
```diff
@@ -129,39 +137,45 @@ static GstMemoryFastMalloc* gstAllocatorFastMallocMemShare(GstMemoryFastMalloc*
+
+    auto other = GST_MEMORY_FAST_MALLOC_CAST(otherMemory);
+    auto otherSpan = other->data.subspan(other->base.offset);
+    return thisSpan.data() == otherSpan.data();
```
This is not correct. The original formula was:
```c
return memory->data + memory->base.offset + memory->base.size == other->data + other->base.offset;
```
However, this is equivalent to:
```c
return memory->data + memory->base.offset == other->data + other->base.offset;
```

This is an insidious bug because WebKit will run fine in your computer, but you will have lost zero-copy on elements handling chunked data like demuxers.

Instead you can do:

```cpp
auto visibleSpan1 = memory1->data.subspan(memory1->base.offset, memory1->base.size);
auto visibleSpan2 = memory2->data.subspan(memory2->base.offset, memory2->base.size);
return visibleSpan1.end() == visibleSpan2.begin();
```

ntrrgc reviewed (changes requested):
I have learned a lot about GstMemory and GstAllocator today. I do now understand GstAllocatorFastMalloc.

I fixed the problems I explained in the review and added more documentation in places.

```diff
diff --git a/Source/WTF/wtf/MallocSpan.h b/Source/WTF/wtf/MallocSpan.h
index b0973554d68b..d22270e3e075 100644
--- a/Source/WTF/wtf/MallocSpan.h
+++ b/Source/WTF/wtf/MallocSpan.h
@@ -128,6 +128,14 @@ public:
         return MallocSpan { static_cast<T*>(ptr), sizeInBytes };
     }
 
+    static MallocSpan tryAlignedMalloc(size_t alignment, size_t sizeInBytes)
+    {
+        auto* ptr = Malloc::tryAlignedMalloc(alignment, sizeInBytes);
+        if (!ptr)
+            return { };
+        return MallocSpan { static_cast<T*>(ptr), sizeInBytes };
+    }
+
     void realloc(size_t newSizeInBytes)
     {
         RELEASE_ASSERT(!(newSizeInBytes % sizeof(T)));
diff --git a/Source/WebCore/platform/graphics/gstreamer/GstAllocatorFastMalloc.cpp b/Source/WebCore/platform/graphics/gstreamer/GstAllocatorFastMalloc.cpp
index 255a15ac63cb..7c2084e7496a 100644
--- a/Source/WebCore/platform/graphics/gstreamer/GstAllocatorFastMalloc.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/GstAllocatorFastMalloc.cpp
@@ -21,13 +21,23 @@
 #include "GstAllocatorFastMalloc.h"
 
 #include <gst/gst.h>
+#include <wtf/FastMalloc.h>
+#include <wtf/MallocSpan.h>
 #include <wtf/StdLibExtras.h>
 #include <wtf/glib/WTFGType.h>
 
 typedef struct {
+    // Instance of the base GstMemory class (we're effectively subclassing it, not unlike GstMemorySystem).
     GstMemory base;
+    // Span to the tail of an GstMemoryFastMalloc allocation block where the user memory contents will be stored.
+    // gstMemoryFastMallocNew() will allocate enough space to fit such tail after GstMemoryFastMalloc.
+    // gstAllocatorFastMallocMemShare() only allocates enough space for the GstMemoryFastMalloc struct and
+    // instead shares the `data` span with the original (parent) GstMemoryFastMalloc.
     std::span<uint8_t> data;
 } GstMemoryFastMalloc;
+//
+// Note: when allocating GstMemoryFastMalloc we also allocate a long enough tail to hold the contents
+// the user will store. That tail is what will be returned by our map() function.
 
 typedef struct {
 } GstAllocatorFastMallocPrivate;
@@ -48,38 +58,41 @@ WEBKIT_DEFINE_TYPE(GstAllocatorFastMalloc, gst_allocator_fast_malloc, GST_TYPE_A
 
 static GstMemoryFastMalloc* gstMemoryFastMallocNew(GstAllocator* allocator, gsize size, gsize alignment, gsize offset, gsize padding, GstMemoryFlags flags)
 {
-    // alignment should be a (power-of-two - 1).
+    ASSERT(G_TYPE_CHECK_INSTANCE_TYPE(allocator, gst_allocator_fast_malloc_get_type()));
+
+    // Alignment should be a (power-of-two - 1). That is GStreamer's convention.
     alignment |= gst_memory_alignment;
     ASSERT(!((alignment + 1) & alignment));
 
     // GStreamer's allocator requires heap allocations.
     DisableMallocRestrictionsForCurrentThreadScope disableMallocRestrictions;
 
-    gsize headerSize = (sizeof(GstMemoryFastMalloc) + alignment) & ~alignment;
-    gsize allocationSize = offset + size + padding;
+    // The block we get from FastMalloc consists of:
+    // (1) *Header*: GstMemoryFastMalloc struct, plus any extra space needed to satisfy alignment.
+    gsize headerSize = (sizeof(GstMemoryFastMalloc) + alignment) & ~alignment; // Rounds up to the nearest alignment unit.
+    // (2) *Tail*: the memory that is accessible to the user via GstMemory.map().
+    gsize tailSize = offset + size + padding;
 
-    auto totalSize = headerSize + allocationSize;
-    auto memory = static_cast<GstMemoryFastMalloc*>(tryFastAlignedMalloc(alignment + 1, totalSize));
-    if (!memory)
+    auto totalSize = headerSize + tailSize;
+    auto storage = MallocSpan<uint8_t, FastAlignedMalloc>::tryAlignedMalloc(alignment + 1 /* Power of 2 */, totalSize);
+    if (!storage)
         return nullptr;
+    auto wholeAllocation = storage.leakSpan();
 
-    auto data = unsafeMakeSpan(reinterpret_cast<uint8_t*>(memory), totalSize);
-    memory->data = data.subspan(headerSize);
+    auto& header = WTF::reinterpretCastSpanStartTo<GstMemoryFastMalloc>(wholeAllocation);
+    header.data = wholeAllocation.subspan(headerSize);
 
-    auto span = memory->data.subspan(0, size);
     if (offset && (flags & GST_MEMORY_FLAG_ZERO_PREFIXED))
-        memsetSpan(span.subspan(0, offset), 0);
+        memsetSpan(header.data.subspan(0, offset), 0);
     if (padding && (flags & GST_MEMORY_FLAG_ZERO_PADDED))
-        memsetSpan(span.subspan(offset + size, padding), 0);
+        memsetSpan(header.data.subspan(offset + size, padding), 0);
 
-    gst_memory_init(GST_MEMORY_CAST(memory), flags, allocator, nullptr, allocationSize, alignment, offset, size);
-    return memory;
+    gst_memory_init(&header.base, flags, allocator, nullptr, tailSize, alignment, offset, size);
+    return &header;
 }
 
 static GstMemory* gstAllocatorFastMallocAlloc(GstAllocator* allocator, gsize size, GstAllocationParams* params)
 {
-    ASSERT(G_TYPE_CHECK_INSTANCE_TYPE(allocator, gst_allocator_fast_malloc_get_type()));
-
     return GST_MEMORY_CAST(gstMemoryFastMallocNew(allocator, size, params->align, params->prefix, params->padding, params->flags));
 }
 
@@ -140,20 +153,30 @@ static GstMemory* gstAllocatorFastMallocMemShare(GstMemory* baseMemory, gssize o
     return GST_MEMORY_CAST(sharedMemory);
 }
 
-static gboolean gstAllocatorFastMallocMemIsSpan(GstMemory* baseMemory, GstMemory* otherMemory, gsize* offset)
+// Will be called by gst_memory_is_span() after checking that both GstMemories have
+// the same allocators and they both share the same non-null parent.
+//
+// Returns whether the visible end of the first memory coincides with the visible
+// start of the second memory and fills `offset` with the offset of the first memory
+// within the visible span of the parent memory.
+//
+// This is used to avoid expensive concatenation: If the function returns true, it
+// means you can use gst_memory_share() on the parent memory with the filled offset
+// and the combined size of both memories to get a GstMemory that spans both previous
+// memories without doing any copies. See _sysmem_is_span() in GStreamer for reference.
+static gboolean gstAllocatorFastMallocMemIsSpan(GstMemory* _memory1, GstMemory* _memory2, gsize* offset)
 {
-    auto memory = GST_MEMORY_FAST_MALLOC_CAST(baseMemory);
+    auto memory1 = GST_MEMORY_FAST_MALLOC_CAST(_memory1);
+    auto memory2 = GST_MEMORY_FAST_MALLOC_CAST(_memory2);
     if (offset) {
-        auto parent = GST_MEMORY_FAST_MALLOC_CAST(memory->base.parent);
+        auto parent = GST_MEMORY_FAST_MALLOC_CAST(memory1->base.parent);
         ASSERT(parent);
-        *offset = memory->base.offset - parent->base.offset;
+        *offset = memory1->base.offset - parent->base.offset;
     }
 
-    auto thisSpan = memory->data.subspan(memory->base.offset, memory->base.size);
-
-    auto other = GST_MEMORY_FAST_MALLOC_CAST(otherMemory);
-    auto otherSpan = other->data.subspan(other->base.offset);
-    return thisSpan.data() == otherSpan.data();
+    auto visibleSpan1 = memory1->data.subspan(memory1->base.offset, memory1->base.size);
+    auto visibleSpan2 = memory2->data.subspan(memory2->base.offset, memory2->base.size);
+    return visibleSpan1.end() == visibleSpan2.begin();
 }
 
 static void gstAllocatorFastMallocConstructed(GObject* object)
```

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GstAllocatorFastMalloc.cpp:
```diff
@@ -129,39 +150,55 @@ static GstMemoryFastMalloc* gstAllocatorFastMallocMemShare(GstMemoryFastMalloc*
-    gstAllocatorClass->alloc = gstAllocatorFastMallocAlloc;
-    gstAllocatorClass->free = gstAllocatorFastMallocFree;
+    G_OBJECT_CLASS(gst_allocator_fast_malloc_parent_class)->constructed(object);
+    GST_OBJECT_FLAG_SET(GST_OBJECT_CAST(object), GST_OBJECT_FLAG_MAY_BE_LEAKED);
```
Why?

philn replied:
It's static, see 271861@main.

calvaris reviewed (approved):
LGTM but my knowledge of this is shallow.

philn committed:
[GStreamer] Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in FastMalloc allocator
https://bugs.webkit.org/show_bug.cgi?id=285528

Reviewed by Xabier Rodriguez-Calvar.

The allocator was updated to modern WebKit/GObject code style and now stores allocated data in a
span.

Co-authored-by: Alicia Boya Garcia <aboya@igalia.com>

* Source/WTF/wtf/MallocSpan.h:
(WTF::MallocSpan::tryAlignedMalloc):
* Source/WebCore/platform/graphics/gstreamer/GstAllocatorFastMalloc.cpp:
(gstMemoryFastMallocNew):
(gstAllocatorFastMallocAlloc):
(gstAllocatorFastMallocMemMap):
(gstAllocatorFastMallocMemUnmap):
(gstAllocatorFastMallocMemCopy):
(gstAllocatorFastMallocMemShare):
(gstAllocatorFastMallocMemIsSpan):
(gstAllocatorFastMallocConstructed):
(gst_allocator_fast_malloc_class_init):
(gst_allocator_fast_malloc_init): Deleted.
* Tools/Scripts/webkitpy/style/checker.py:

Canonical link: https://commits.webkit.org/288928@main
